### PR TITLE
TopologyAwareHints: Take lock in HasPopulatedHints

### LINKED ISSUE #118188: Topology cache's HasPopulatedHints method can attempt concurrent map access
### What happened?

I noticed that kube-controller-manager terminated with a concurrent map access:
```
fatal error: concurrent map read and map write

goroutine 2456 [running]:
k8s.io/apimachinery/pkg/util/sets.Set[...].Has(0x1?, {0xc00507d650?, 0x23?})
	vendor/k8s.io/apimachinery/pkg/util/sets/set.go:83 +0x2b
k8s.io/kubernetes/pkg/controller/endpointslice/topologycache.(*TopologyCache).HasPopulatedHints(...)
	pkg/controller/endpointslice/topologycache/topologycache.go:266
k8s.io/kubernetes/pkg/controller/endpointslice.(*reconciler).reconcileByAddressType(0xc000724d20, 0xc002f5a698, {0xc0048bdec8, 0x1, 0x0?}, {0xc0048bded8, 0x1, 0x1}, {0x0, 0x0, ...}, ...)
	pkg/controller/endpointslice/reconciler.go:268 +0x17d4
k8s.io/kubernetes/pkg/controller/endpointslice.(*reconciler).reconcile(0xc000724d20, 0xc002f5a698, {0xc0048bdec8, 0x1, 0x1}, {0xc0048bded0, 0x1, 0xc00938fbc8?}, {0x0, 0x0, ...})
	pkg/controller/endpointslice/reconciler.go:108 +0x625
k8s.io/kubernetes/pkg/controller/endpointslice.(*Controller).syncService(0xc00028b5e0, {0xc008935f80, 0x26})
	pkg/controller/endpointslice/endpointslice_controller.go:382 +0xaea
k8s.io/kubernetes/pkg/controller/endpointslice.(*Controller).processNextWorkItem(0xc00028b5e0)
	pkg/controller/endpointslice/endpointslice_controller.go:282 +0xd6
k8s.io/kubernetes/pkg/controller/endpointslice.(*Controller).worker(...)
	pkg/controller/endpointslice/endpointslice_controller.go:271
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
	vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x6975622e79727473?, {0x4e48440, 0xc003d47860}, 0x1, 0xc000cdbda0)
	vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x6462333331386365?, 0x3b9aca00, 0x0, 0x65?, 0x3335343863633839?)
	vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(0x69632e3130646c69?, 0x6968736e65706f2e?, 0x632f67726f2e7466?)
	vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:161 +0x25
created by k8s.io/kubernetes/pkg/controller/endpointslice.(*Controller).Run
	pkg/controller/endpointslice/endpointslice_controller.go:260 +0x32c
```

### What did you expect to happen?

No concurrent map access, and no fatal error.

### How can we reproduce it (as minimally and precisely as possible)?

No reproducer.  I noticed this in an internal CI run.  

### Anything else we need to know?

This appears to be an obvious (in retrospect) logic error with a straightforward solution, so I assume that it isn't necessary to provide a reproducer.  

### Kubernetes version

This is from an OpenShift cluster based on Kubernetes 1.27.
```
I0504 18:49:02.405503       1 controllermanager.go:203] "Starting" version="v1.27.0-rc.0.3023+8932ffc9dd4df0-dirty"
```

### Cloud provider

AWS.

### OS version

Red Hat Enterprise Linux CoreOS 414.92.202305040056-0.

### Install tools

_No response_

### Container runtime (CRI) and version (if applicable)

_No response_

### Related plugins (CNI, CSI, ...) and versions (if applicable)

_No response_

### PR DESCRIPTION
**Pull request by Miciah**
-------

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Prevent potential concurrent map access by taking a lock before reading the topology cache's `hintsPopulatedByService` map.

* `staging/src/k8s.io/endpointslice/topologycache/topologycache.go` (`setHintsLocked`, `hasPopulatedHintsLocked`): New helper functions.  These are the same as the existing `SetHints` and `HasPopulatedHints` methods except that these helpers assume that a lock is already held.
(`SetHints`): Use `setHintsLocked`.
(`HasPopulatedHints`): Take a lock and use `hasPopulatedHintsLocked`.
(`AddHints`): Take a lock and use `setHintsLocked` and `hasPopulatedHintsLocked`.
* `staging/src/k8s.io/endpointslice/topologycache/topologycache_test.go` (`TestTopologyCacheRace`): Add a goroutine that calls `HasPopulatedHints`.

#### Which issue(s) this PR fixes:

Fixes #118188.

#### Does this PR introduce a user-facing change?

```release-note
Fix a concurrent map access in TopologyCache's `HasPopulatedHints` method.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by aojea** [comment]
was this not fixed by https://github.com/kubernetes/kubernetes/pull/117249?

if not, can we add a test like the one I suggested in https://github.com/kubernetes/kubernetes/pull/117249#issuecomment-1505636060 to cover the regression?

**Comment by robscott** [comment]
Thanks @Miciah!

/ok-to-test

**Comment by atiratree** [comment]
> It was not, but thanks for pointing me to that change! I was able to reproduce the race condition by modifying TestTopologyCacheRace.

maybe it would be good to increase the chance by calling `SetHints` directly or add a few more goroutines?

> > if not, can we add a test like the one I suggested in https://github.com/kubernetes/kubernetes/pull/117249#issuecomment-1505636060 to cover the regression?

> Is modifying TestTopologyCacheRace sufficient, or do you want a separate test?

@aojea

**Comment by aojea** [comment]
modifying it LGTM, duplicating the test too ... is just to avoid regressions

**Comment by aojea** [comment]
I mean, modifying if it covers both races, otherwise just duplicate it

File: pkg/controller/endpointslice/topologycache/topologycache.go [hunk 1]
```diff
@@ -263,6 +269,13 @@ func (t *TopologyCache) SetNodes(nodes []*v1.Node) {
 
 // HasPopulatedHints checks whether there are populated hints for a given service in the cache.
 func (t *TopologyCache) HasPopulatedHints(serviceKey string) bool {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	return t.hasPopulatedHintsLocked(serviceKey)
+}
+
+func (t *TopologyCache) hasPopulatedHintsLocked(serviceKey string) bool {
```
Comment (atiratree) [inline pkg/controller/endpointslice/topologycache/topologycache.go:278]: the naming might be a bit misleading, what about `hasPopulatedHintsWithoutLock` or `hasPopulatedHintsThreadUnsafe`?

File: pkg/controller/endpointslice/topologycache/topologycache.go [hunk 2]
```diff
@@ -175,6 +177,10 @@ func (t *TopologyCache) SetHints(serviceKey string, addrType discovery.AddressTy
 	t.lock.Lock()
 	defer t.lock.Unlock()
 
+	t.setHintsLocked(serviceKey, addrType, allocatedHintsByZone)
+}
+
+func (t *TopologyCache) setHintsLocked(serviceKey string, addrType discovery.AddressType, allocatedHintsByZone EndpointZoneInfo) {
```
Comment (atiratree) [inline pkg/controller/endpointslice/topologycache/topologycache.go:183]: +here

File: pkg/controller/endpointslice/topologycache/topologycache.go (hunk 1 above)
Comment (Miciah) [reply pkg/controller/endpointslice/topologycache/topologycache.go:278]: Are you sure?  I see the naming pattern of `foo` (or `Foo`) that takes a lock and `fooLocked` that assumes a lock is held all over the Kubernetes codebase when I grep for `func .*Locked(`.  Here's one example out of dozens: https://github.com/kubernetes/kubernetes/blob/decf1e1a9b53a4f4b24bb780d6433e2d5e959d83/staging/src/k8s.io/client-go/tools/cache/heap.go#L188-L216

I see one match for `func.*WithoutLock`, but it actually takes a lock: https://github.com/kubernetes/kubernetes/blob/decf1e1a9b53a4f4b24bb780d6433e2d5e959d83/vendor/github.com/peterbourgon/diskv/diskv.go#L546-L551

I see only a couple matches for `func.*ThreadUnsafe`, e.g.: https://github.com/kubernetes/kubernetes/blob/decf1e1a9b53a4f4b24bb780d6433e2d5e959d83/pkg/kubelet/util/manager/watch_based_manager.go#L66-L82

In conclusion:

* `fooWithoutLock` seems to mean the opposite of what we want here. 
* `fooThreadUnsafe` is unusual but has precedent.
* `fooLocked` seems to be the prevailing pattern.  

However, if you strongly prefer "hasPopulatedHintsThreadUnsafe", I'll make the change.  Let me know if you indeed want me to do that.

**Comment by atiratree** [comment]
@Miciah can you please rebase the PR?

**Comment by atiratree** [comment]
seems this is a flake
/test pull-kubernetes-e2e-gce

**Comment by robscott** [comment]
Thanks @Miciah! Would like to get another set of eyes on this for final LGTM in case I'm missing anything here.

/cc @aojea @swetharepakula 
/approve
